### DISCUSSION TITLE
How to Require WebAuthN on top of IdP

### DISCUSSION THREAD

**Discussion by CollinEstes** [keycloak/keycloak#22357] [Q&A]
I'm having trouble figuring out an authentication flow configuration that would do the following:

1. Users login with external IdP assigned to realm
2. First login auto-provisions new user, and  requires the new user to register a webAuthN passkey. 
3. Subsequent logins have the user first redirected to sign in with their external IdP, then are also required to authenticate with their previously setup webAuthN authenticator. 

I have not been able to achieve the above, either my setup works on subsequent logins and breaks the new user flow, or I can have the new user flow working as intended but can't seem to require the subsequent logins to require both the IdP auth then the passkey auth. 

What I have tried so far:

1. Registration setting for new users to register webAuthN and a "post login" flow on the Identity Provider with a custom authentication flow that has a single task of "WebAuthN passwordless authenticator" as required. 

- This only works for users that are already in the system, and who went through the initial login before I added the "post login" flow to the IdP. New users don't get to the registration before being stopped as requiring an authenticator. My guess is the "post login" flow happens before the new user registration process.
- I tried changing the post flow authentication flow to be "conditional - user authenticator" and then in that sub flow having the WebAuthN task but that doesn't seem to work. 

2. Registration setting for new users to register webAuthN and a custom "browser flow" with a "WebAuthN passwordless authenticator" step set as required and the identity provider redirect still the same as the normal "Alternative"

- The registration works as intended for first time users, subsequent users are immediately promoted to enter their security key, after which they are allowed through. So it bypasses the IdP login and I need them to do both. 

3. Registration setting for new users to register webAuthN and a custom "browser flow" with a "WebAuthN passwordless authenticator" step set as required and the identity provider redirect still the same as the normal "Required". 

- The registration works as intended for first time users, subsequent users are prompted to do the IdP redirect but on the callback Keycloak errors with an "invalid username and password" message. If I just take the standard browser flow and make that "Identity Provider redirect" "Required" I get the same behavior so clearly that isn't right. 

I cant seem to be able to setup the right configurations here to achieve what I'm looking for. The core problem I'm trying to solve is I want to use external IdP but I need to layer on top a stronger authenticator as I can't assume authenticator strength with the IdP alone. 

Really appreciate any help anyone can. Give me here.  Thanks! 🙏

**Comment by cgeorgilakis** [comment]
Unfortunately, Authentication Flows are not executed after Identity Provider User log in/ Identity Provider Redirect. See [the related jira issue](https://github.com/keycloak/keycloak/issues/10250)for details. We have implemented a [PR](https://github.com/keycloak/keycloak/pull/12891) trying to solve it without any review yet.

The only possible solution with Keycloak core is 1 with the drawback that authendication happens on Identity Provider level not in client level. Maybe if you change First Login Flow, it will work this scenario.
